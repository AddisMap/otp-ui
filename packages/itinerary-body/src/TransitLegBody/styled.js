import styled from "styled-components";

import { TransparentButton } from "../styled-legs";

export const AgencyInfo = styled.div`
  margin-top: 5px;

  a {
    color: #337ab7;
    text-decoration: none;
  }

  img {
    vertical-align: middle;
  }
`;

export const IntermediateStops = styled.div`
  display: block;
`;

export const StopMarker = styled.div`
  float: left;
  margin-left: -36px;
  color: #fff;
`;

export const StopName = styled.div`
  color: #999;
  font-size: 14px;
  margin-top: 3px;
`;

export const StopRow = styled.div`
  z-index: 30;
  position: relative;
`;

export const TransitAlert = styled.div`
  margin-top: 5px;
  background-color: #eee;
  padding: 8px;
  color: #000;
  border-radius: 4px;
`;

export const TransitAlertBody = styled.div`
  font-size: 12px;
  margin-left: 30px;
  white-space: pre-wrap;
`;

export const TransitAlertEffectiveDate = styled.div`
  margin-top: 5px;
  margin-left: 30px;
  font-size: 12px;
  font-style: italic;
`;

export const TransitAlertHeader = styled.div`
  font-size: 14px;
  margin-left: 30px;
  font-weight: 600;
`;

export const TransitAlertIconContainer = styled.div`
  float: left;
  font-size: 18px;
`;

export const TransitAlerts = styled.div`
  display: block;
  margin-top: 3px;
`;

export const TransitAlertToggle = styled(TransparentButton)`
  display: inline-block;
  margin-top: 8px;
  color: #d14727;
  font-weight: 400;
  cursor: pointer;
`;

export const ViewTripButton = styled(TransparentButton)`
  color: #008;
  outline: none;
  height: 14px;
  padding-top: 0;
  line-height: 1;
  margin-left: 5px;
  border-left: 1px solid #000;
`;
