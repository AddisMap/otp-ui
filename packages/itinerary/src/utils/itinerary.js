export const transitModes = [
  "TRAM",
  "BUS",
  "SUBWAY",
  "FERRY",
  "RAIL",
  "GONDOLA"
];

export function isTransit(mode) {
  return transitModes.includes(mode) || mode === "TRANSIT";
}

// TODO: Is this a place where we can get these colors dynamically?
export function getMapColor(mode) {
  switch (mode) {
    case "WALK":
      return "#444";
    case "BICYCLE":
      return "#0073e5";
    case "SUBWAY":
      return "#f00";
    case "RAIL":
      return "#b00";
    case "BUS":
      return "#080";
    case "TRAM":
      return "#800";
    case "FERRY":
      return "#008";
    case "CAR":
      return "#444";
    case "MICROMOBILITY":
      return "#f5a729";
    default:
      return "#aaa";
  }
}

// TODO: temporary code; handle via migrated OTP i18n language table
export function getStepDirection(step) {
  switch (step.relativeDirection) {
    case "DEPART":
      return `Head ${step.absoluteDirection.toLowerCase()}`;
    case "LEFT":
      return "Left";
    case "HARD_LEFT":
      return "Hard left";
    case "SLIGHTLY_LEFT":
      return "Slight left";
    case "CONTINUE":
      return "Continue";
    case "SLIGHTLY_RIGHT":
      return "Slight right";
    case "RIGHT":
      return "Right";
    case "HARD_RIGHT":
      return "Hard right";
    case "CIRCLE_CLOCKWISE":
      return "Follow circle clockwise";
    case "CIRCLE_COUNTERCLOCKWISE":
      return "Follow circle counterclockwise";
    case "ELEVATOR":
      return "Take elevator";
    case "UTURN_LEFT":
      return "Left U-turn";
    case "UTURN_RIGHT":
      return "Right U-turn";
    default:
      return step.relativeDirection;
  }
}

export function getStepInstructions(step) {
  const conjunction = step.relativeDirection === "ELEVATOR" ? "to" : "on";
  return `${getStepDirection(step)} ${conjunction} ${step.streetName}`;
}

export function getStepStreetName(step) {
  if (step.streetName === "road") return "Unnamed Road";
  if (step.streetName === "path") return "Unnamed Path";
  return step.streetName;
}

/**
 * Get the configured company object for the given network string if the company
 * has been defined in the provided companies array config.
 */
function getCompanyForNetwork(networkString, companies = []) {
  const company = companies.find(co => co.id === networkString);
  if (!company) {
    console.warn(
      `No company found in config.yml that matches rented vehicle network: ${networkString}`,
      companies
    );
  }
  return company;
}

/**
 * Returns mode name by checking the vertex type (VertexType class in OTP) for
 * the provided place. NOTE: this is currently only intended for vehicles at
 * the moment (not transit or walking).
 *
 * TODO: I18N
 * @param  {string} place place from itinerary leg
 */
export function getModeForPlace(place) {
  switch (place.vertexType) {
    case "CARSHARE":
      return "car";
    case "VEHICLERENTAL":
      return "E-scooter";
    // TODO: Should the type change depending on bike vertex type?
    case "BIKESHARE":
    case "BIKEPARK":
      return "bike";
    // If company offers more than one mode, default to `vehicle` string.
    default:
      return "vehicle";
  }
}

export function getPlaceName(place, companies) {
  // If address is provided (i.e. for carshare station, use it)
  if (place.address) return place.address.split(",")[0];
  if (place.networks && place.vertexType === "VEHICLERENTAL") {
    // For vehicle rental pick up, do not use the place name. Rather, use
    // company name + vehicle type (e.g., SPIN E-scooter). Place name is often just
    // a UUID that has no relevance to the actual vehicle. For bikeshare, however,
    // there are often hubs or bikes that have relevant names to the user.
    const company = getCompanyForNetwork(place.networks[0], companies);
    if (company) {
      return `${company.label} ${getModeForPlace(place)}`;
    }
  }
  // Default to place name
  return place.name;
}

export function toSentenceCase(str) {
  if (str == null) {
    return "";
  }
  const typedString = String(str);
  return (
    typedString.charAt(0).toUpperCase() + typedString.substr(1).toLowerCase()
  );
}

export function getLegModeLabel(leg) {
  switch (leg.mode) {
    case "BICYCLE_RENT":
      return "Biketown";
    case "CAR":
      return leg.hailedCar ? "Ride" : "Drive";
    case "GONDOLA":
      return "Aerial Tram";
    case "TRAM":
      if (leg.routeLongName.toLowerCase().indexOf("streetcar") !== -1)
        return "Streetcar";
      return "Light Rail";
    case "MICROMOBILITY":
      return "Ride";
    default:
      return toSentenceCase(leg.mode);
  }
}

/** Looks up an operator from the provided configuration */
export const getOperatorFromConfig = (id, config) =>
  config.operators.find(operator => operator.id === id) || null;
