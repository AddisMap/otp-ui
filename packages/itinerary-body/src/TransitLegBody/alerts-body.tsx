import moment from "moment";
import { Alert } from "@opentripplanner/types";
import React, { FunctionComponent, ReactElement } from "react";
import { FormattedMessage } from "react-intl";

import * as S from "../styled";

interface Props {
  alerts: Alert[];
  AlertIcon?: FunctionComponent;
}

export default function AlertsBody({
  alerts,
  AlertIcon = S.DefaultAlertBodyIcon
}: Props): ReactElement {
  return (
    <S.TransitAlerts>
      {alerts
        .sort((a, b) => b.effectiveStartDate - a.effectiveStartDate)
        .map((alert, i) => {
          // If alert is effective as of +/- one day, use today, tomorrow, or
          // yesterday with time. Otherwise, use long date format.
          // FIXME: adding 24 hours to a date that is "today" can lead moment to
          // report the result to still be "today" (see OTP-RR story).
          const dayDiff = moment(alert.effectiveStartDate).diff(
            moment(),
            "days"
          );
          return (
            <S.TransitAlert key={i} href={alert.alertUrl}>
              <S.TransitAlertIconContainer>
                <AlertIcon />
              </S.TransitAlertIconContainer>
              {alert.alertHeaderText ? (
                <S.TransitAlertHeader>
                  {alert.alertHeaderText}
                </S.TransitAlertHeader>
              ) : null}
              <S.TransitAlertBody>
                {alert.alertDescriptionText}
              </S.TransitAlertBody>
              <S.TransitAlertEffectiveDate>
                <FormattedMessage
                  defaultMessage={"{dateTime, date, long}"}
                  description="Text describing when an alert takes effect"
                  id="otpUi.TransitLegBody.AlertsBody.effectiveDate"
                  values={{
                    dateTime: alert.effectiveStartDate,
                    dayDiff,
                    includeTime: Math.abs(dayDiff) <= 1
                  }}
                />
              </S.TransitAlertEffectiveDate>
            </S.TransitAlert>
          );
        })}
    </S.TransitAlerts>
  );
}
