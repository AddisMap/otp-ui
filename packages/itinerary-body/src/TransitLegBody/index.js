import moment from "moment";
import {
  legType,
  transitOperatorType
} from "@opentripplanner/core-utils/lib/types";
import PropTypes from "prop-types";
import React, { Component } from "react";
import { ExclamationTriangle } from "styled-icons/fa-solid";
import { VelocityTransitionGroup } from "velocity-react";

import { formatDuration } from "@opentripplanner/core-utils/lib/time";

import * as Styled from "../styled";
import ViewTripButton from "./view-trip-button";

// TODO use pluralize that for internationalization (and complex plurals, i.e., not just adding 's')
function pluralize(str, list) {
  return `${str}${list.length > 1 ? "s" : ""}`;
}

function DefaultTransitLegSummary({ leg, stopsExpanded }) {
  return (
    <>
      {leg.duration && <span>Ride {formatDuration(leg.duration)}</span>}
      {leg.intermediateStops && (
        <span>
          {" / "}
          {leg.intermediateStops.length + 1}
          {" stops "}
          <Styled.CaretToggle expanded={stopsExpanded} />
        </span>
      )}
    </>
  );
}

DefaultTransitLegSummary.propTypes = {
  leg: legType.isRequired,
  stopsExpanded: PropTypes.bool.isRequired
};

function DefaultRouteDescription({ leg }) {
  const { headsign, routeLongName, routeShortName } = leg;
  return (
    <Styled.LegDescriptionForTransit>
      {routeShortName && (
        <div>
          <Styled.LegDescriptionRouteShortName>
            {routeShortName}
          </Styled.LegDescriptionRouteShortName>
        </div>
      )}
      <Styled.LegDescriptionRouteLongName>
        {routeLongName}
        {headsign && (
          <span>
            <Styled.LegDescriptionHeadsignPrefix>
              {" to "}
            </Styled.LegDescriptionHeadsignPrefix>
            {headsign}
          </span>
        )}
      </Styled.LegDescriptionRouteLongName>
    </Styled.LegDescriptionForTransit>
  );
}

DefaultRouteDescription.propTypes = {
  leg: legType.isRequired
};

class TransitLegBody extends Component {
  constructor(props) {
    super(props);
    this.state = {
      alertsExpanded: false,
      stopsExpanded: false
    };
  }

  onToggleStopsClick = () => {
    const { stopsExpanded } = this.state;
    this.setState({ stopsExpanded: !stopsExpanded });
  };

  onToggleAlertsClick = () => {
    const { alertsExpanded } = this.state;
    this.setState({ alertsExpanded: !alertsExpanded });
  };

  onSummaryClick = () => {
    const { leg, legIndex, setActiveLeg } = this.props;
    setActiveLeg(legIndex, leg);
  };

  render() {
    const {
      leg,
      longDateFormat,
      RouteDescription,
      setViewedTrip,
      showAgencyInfo,
      timeFormat,
      TransitLegSummary,
      transitOperator
    } = this.props;
    const { agencyBrandingUrl, agencyName, agencyUrl, alerts } = leg;
    const { alertsExpanded, stopsExpanded } = this.state;

    // If the config contains an operator with a logo URL, prefer that over the
    // one provided by OTP (which is derived from agency.txt#agency_branding_url)
    const logoUrl =
      transitOperator && transitOperator.logo
        ? transitOperator.logo
        : agencyBrandingUrl;

    const expandAlerts =
      alertsExpanded || (leg.alerts && leg.alerts.length < 3);

    return (
      <Styled.LegBody>
        {/* The Route Icon/Name Bar; clickable to set as active leg */}
        <Styled.LegClickable onClick={this.onSummaryClick}>
          <RouteDescription leg={leg} transitOperator={transitOperator} />
        </Styled.LegClickable>

        {/* Agency information */}
        {showAgencyInfo && (
          <Styled.AgencyInfo>
            Service operated by{" "}
            <a href={agencyUrl} rel="noopener noreferrer" target="_blank">
              {agencyName}
              {logoUrl && (
                <img alt={`${agencyName} logo`} src={logoUrl} height={25} />
              )}
            </a>
          </Styled.AgencyInfo>
        )}

        {/* Alerts toggle */}
        {alerts && alerts.length > 2 && (
          <Styled.TransitAlertToggle onClick={this.onToggleAlertsClick}>
            <ExclamationTriangle size={15} /> {alerts.length}{" "}
            {pluralize("alert", alerts)}{" "}
            <Styled.CaretToggle expanded={alertsExpanded} />
          </Styled.TransitAlertToggle>
        )}

        {/* The Alerts body, if visible */}
        <VelocityTransitionGroup
          enter={{ animation: "slideDown" }}
          leave={{ animation: "slideUp" }}
        >
          {expandAlerts && (
            <AlertsBody
              alerts={leg.alerts}
              longDateFormat={longDateFormat}
              timeFormat={timeFormat}
            />
          )}
        </VelocityTransitionGroup>
        {/* The "Ride X Min / X Stops" Row, including IntermediateStops body */}
        {leg.intermediateStops && leg.intermediateStops.length > 0 && (
          <Styled.TransitLegDetails>
            {/* The header summary row, clickable to expand intermediate stops */}
            <Styled.TransitLegDetailsHeader onClick={this.onToggleStopsClick}>
              <TransitLegSummary leg={leg} stopsExpanded={stopsExpanded} />

              {/* The ViewTripButton. TODO: make configurable */}
              <ViewTripButton
                tripId={leg.tripId}
                fromIndex={leg.from.stopIndex}
                setViewedTrip={setViewedTrip}
                toIndex={leg.to.stopIndex}
              />
            </Styled.TransitLegDetailsHeader>
            {/* IntermediateStops expanded body */}
            <VelocityTransitionGroup
              enter={{ animation: "slideDown" }}
              leave={{ animation: "slideUp" }}
            >
              {stopsExpanded ? (
                <IntermediateStops stops={leg.intermediateStops} />
              ) : null}
            </VelocityTransitionGroup>

            {/* Average wait details, if present */}
            {leg.averageWait && (
              <span>Typical Wait: {formatDuration(leg.averageWait)}</span>
            )}
          </Styled.TransitLegDetails>
        )}
      </Styled.LegBody>
    );
  }
}

TransitLegBody.propTypes = {
  leg: legType.isRequired,
  legIndex: PropTypes.number.isRequired,
  longDateFormat: PropTypes.string.isRequired,
  RouteDescription: PropTypes.elementType,
  setActiveLeg: PropTypes.func.isRequired,
  setViewedTrip: PropTypes.func.isRequired,
  showAgencyInfo: PropTypes.bool.isRequired,
  timeFormat: PropTypes.string.isRequired,
  TransitLegSummary: PropTypes.elementType,
  transitOperator: transitOperatorType
};

TransitLegBody.defaultProps = {
  RouteDescription: DefaultRouteDescription,
  TransitLegSummary: DefaultTransitLegSummary,
  transitOperator: null
};

export default TransitLegBody;

function IntermediateStops({ stops }) {
  return (
    <Styled.IntermediateStops>
      {stops.map((stop, k) => {
        return (
          <Styled.StopRow key={k}>
            <Styled.StopMarker>&bull;</Styled.StopMarker>
            <Styled.StopName>{stop.name}</Styled.StopName>
          </Styled.StopRow>
        );
      })}
    </Styled.IntermediateStops>
  );
}

IntermediateStops.propTypes = {
  stops: PropTypes.arrayOf(PropTypes.shape({})).isRequired
};

function AlertsBody({ alerts, longDateFormat, timeFormat }) {
  return (
    <Styled.TransitAlerts>
      {alerts
        .sort((a, b) => b.effectiveStartDate - a.effectiveStartDate)
        .map((alert, i) => {
          // If alert is effective as of +/- one day, use today, tomorrow, or
          // yesterday with time. Otherwise, use long date format.
          const dateTimeString = moment(alert.effectiveStartDate).calendar(
            null,
            {
              sameDay: `${timeFormat}, [Today]`,
              nextDay: `${timeFormat}, [Tomorrow]`,
              lastDay: `${timeFormat}, [Yesterday]`,
              lastWeek: `${longDateFormat}`,
              sameElse: `${longDateFormat}`
            }
          );
          const effectiveDateString = `Effective as of ${dateTimeString}`;
          return (
            <Styled.TransitAlert key={i} href={alert.alertUrl}>
              <Styled.TransitAlertIconContainer>
                <ExclamationTriangle size={18} />
              </Styled.TransitAlertIconContainer>
              {alert.alertHeaderText ? (
                <Styled.TransitAlertHeader>
                  {alert.alertHeaderText}
                </Styled.TransitAlertHeader>
              ) : null}
              <Styled.TransitAlertBody>
                {alert.alertDescriptionText}
              </Styled.TransitAlertBody>
              <Styled.TransitAlertEffectiveDate>
                {effectiveDateString}
              </Styled.TransitAlertEffectiveDate>
            </Styled.TransitAlert>
          );
        })}
    </Styled.TransitAlerts>
  );
}

AlertsBody.propTypes = {
  alerts: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  longDateFormat: PropTypes.string.isRequired,
  timeFormat: PropTypes.string.isRequired
};
