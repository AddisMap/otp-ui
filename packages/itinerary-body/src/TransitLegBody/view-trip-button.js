import React, { Component } from "react";
import PropTypes from "prop-types";

import * as Styled from "./styled";

class ViewTripButton extends Component {
  onClick = () => {
    const { fromIndex, setViewedTrip, toIndex, tripId } = this.props;
    setViewedTrip({ tripId, fromIndex, toIndex });
  };

  render() {
    return (
      <Styled.ViewTripButton onClick={this.onClick} type="button">
        Trip Viewer
      </Styled.ViewTripButton>
    );
  }
}

ViewTripButton.propTypes = {
  fromIndex: PropTypes.number.isRequired,
  setViewedTrip: PropTypes.func.isRequired,
  tripId: PropTypes.string.isRequired,
  toIndex: PropTypes.number.isRequired
};

export default ViewTripButton;
