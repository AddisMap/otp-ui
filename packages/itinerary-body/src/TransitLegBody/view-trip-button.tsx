import React, { Component, ReactElement } from "react";
import { FormattedMessage } from "react-intl";

import * as S from "../styled";
import { defaultMessages } from "../util";

interface TripSection {
  fromIndex: number;
  toIndex: number;
  tripId: string;
}

export type SetViewedTripFunction = (tripSection: TripSection) => void;

type Props = TripSection & {
  setViewedTrip: SetViewedTripFunction;
};

class ViewTripButton extends Component<Props> {
  onClick = (): void => {
    const { fromIndex, setViewedTrip, toIndex, tripId } = this.props;
    setViewedTrip({ fromIndex, toIndex, tripId });
  };

  render(): ReactElement {
    return (
      <S.ViewerButton onClick={this.onClick} type="button">
        <FormattedMessage
          defaultMessage={defaultMessages["otpUi.TransitLegBody.tripViewer"]}
          description="Link text to the trip viewer"
          id="otpUi.TransitLegBody.tripViewer"
        />
      </S.ViewerButton>
    );
  }
}

export default ViewTripButton;
