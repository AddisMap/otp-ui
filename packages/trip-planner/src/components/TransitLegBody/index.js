import React, { useState } from "react";
import PropTypes from "prop-types";
import * as Styled from "./styled";
import { formatDuration } from "../../utils/time";

const TransitLegBody = ({
  leg,
  legIndex,
  setActiveLeg,
  // Service alerts uses this
  // eslint-disable-next-line no-unused-vars
  longDateFormat,
  operator,
  // Service alerts uses this
  // eslint-disable-next-line no-unused-vars
  timeFormat
}) => {
  const [stopsExpanded, setStopsExpanded] = useState(false);
  // const [alertsExpanded, setAlertsExpanded] = useState(false);
  const {
    agencyBrandingUrl,
    // eslint-disable-next-line no-unused-vars
    agencyName,
    // eslint-disable-next-line no-unused-vars
    agencyUrl,
    // eslint-disable-next-line no-unused-vars
    alerts,
    mode,
    routeShortName,
    routeLongName,
    headsign,
    interlineWithPreviousLeg,
    intermediateStops,
    duration
  } = leg;
  // This is tied to the agency display that is still being decided on
  // eslint-disable-next-line no-unused-vars
  const logoUrl = operator && operator.logo ? operator.logo : agencyBrandingUrl;

  return (
    <Styled.LegBody>
      {/* The Route Icon/Name Bar; clickable to set as active leg */}
      {/* TODO: double check if we need to hide this when interline */}
      {!interlineWithPreviousLeg && (
        <Styled.TransitLegSummary onClick={() => setActiveLeg(legIndex)}>
          <Styled.RouteLongName>
            Board {routeShortName}-{routeLongName}
            {headsign && (
              <>
                {" "}
                <Styled.Lighten>to </Styled.Lighten>
                {headsign}
              </>
            )}
            {/* TODO Need to add the arrival time here with early/late styling EG: Arriving in 9 min */}
          </Styled.RouteLongName>
        </Styled.TransitLegSummary>
      )}

      {/* Agency information */}
      {/* TODO: Marketing is checking to see if they plan show show/hide this. 
        It's ready to go if they want to keep it */}
      {/*
        <Styled.AgencyInfo>
          Service operated by{' '}
          <Styled.AgencyLink href={agencyUrl}>
            {agencyName}
            {logoUrl && (
              <Styled.AgencyLogo src={logoUrl} agencyName={agencyName} />
            )}
          </Styled.AgencyLink>
        </Styled.AgencyInfo>
      */}

      {/* 
        TODO: In the mocks, alerts look to be shown as a list of service alerts 
        while in OTP-RR this is a collapsible list -- need for figure out which direction
        we're going in for this. Have a question into their design team about this
      */}

      {/* Alerts toggle */}
      {/* {alerts && alerts.length > 0 && (
        <div
          onClick={this._onToggleAlertsClick}
          className="transit-alerts-toggle"
        >
          <i className="fa fa-exclamation-triangle" /> {alerts.length}{' '}
          {pluralize('alert', alerts)}{' '}
          <i
            className={`fa fa-caret-${
              this.state.alertsExpanded ? 'up' : 'down'
            }`}
          />
        </div>
      )} */}

      {/* The Alerts body, if visible */}
      {/* <VelocityTransitionGroup
        enter={{ animation: 'slideDown' }}
        leave={{ animation: 'slideUp' }}
      >
        {alertsExpanded && (
          <AlertsBody
            alerts={leg.alerts}
            longDateFormat={longDateFormat}
            timeFormat={timeFormat}
          />
        )}
      </VelocityTransitionGroup> */}
      {/* The "Ride X Min / X Stops" Row, including IntermediateStops body */}
      {intermediateStops && intermediateStops.length > 0 && (
        <Styled.TransitLegDetails>
          <Styled.TransitLegDetailsHeader
            onClick={() => setStopsExpanded(!stopsExpanded)}
          >
            {duration && <>{formatDuration(duration)} ride</>}
            {intermediateStops && (
              <>
                {" ("}
                {intermediateStops.length + 1}
                {" stops) "}
              </>
            )}
            <Styled.ExpandIcon
              expanded={stopsExpanded.toString()}
              mode={mode}
            />
          </Styled.TransitLegDetailsHeader>

          {/* TODO: This does not show up on the mocks. We should ask if trimet
          wants us to carry this feature forward */}
          {/* The ViewTripButton. TODO: make configurable */}
          {/* <ViewTripButton
          //     tripId={leg.tripId}
          //     fromIndex={leg.from.stopIndex}
          //     toIndex={leg.to.stopIndex}
          //   /> */}

          <Styled.IntermediateStopsListWrapper
            expanded={stopsExpanded.toString()}
          >
            <Styled.IntermediateStopsList>
              {stopsExpanded
                ? intermediateStops.map(stop => (
                    <Styled.IntermediateStopRow key={stop.stopId}>
                      {/* TODO: Positioning this stop marker will be tricky */}
                      <Styled.StopMarker>&bull;</Styled.StopMarker>
                      <Styled.StopName>{stop.name}</Styled.StopName>
                    </Styled.IntermediateStopRow>
                  ))
                : null}
            </Styled.IntermediateStopsList>
          </Styled.IntermediateStopsListWrapper>
        </Styled.TransitLegDetails>

        // TODO: What is this and do we want to include it?
        // {/* Average wait details, if present */}
        // {leg.averageWait && (
        //   <span>Typical Wait: {formatDuration(leg.averageWait)}</span>
        // )}
        // </div>
      )}
    </Styled.LegBody>
  );
};

TransitLegBody.propTypes = {
  /** Contains details about leg object that is being displayed */
  leg: PropTypes.shape({
    agencyBrandingUrl: PropTypes.string,
    agencyName: PropTypes.string.isRequired,
    agencyUrl: PropTypes.string.isRequired,
    alerts: PropTypes.arrayOf(PropTypes.shape({}).isRequired),
    routeShortName: PropTypes.string.isRequired,
    routeLongName: PropTypes.string.isRequired,
    headsign: PropTypes.string.isRequired,
    interlineWithPreviousLeg: PropTypes.bool.isRequired,
    intermediateStops: PropTypes.arrayOf(
      PropTypes.shape({
        stopId: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired
      })
    ),
    duration: PropTypes.number.isRequired,
    mode: PropTypes.string.isRequired
  }).isRequired,
  /** The index value of this specific leg within the itinerary */
  legIndex: PropTypes.number.isRequired,
  /** Sets the active leg */
  setActiveLeg: PropTypes.func.isRequired,
  /** The preferred format string for displaying long dates */
  longDateFormat: PropTypes.string.isRequired,
  /** Contains details about the operator of this transit leg */
  operator: PropTypes.shape({
    logo: PropTypes.string
  }),
  /** Contains the preferred format for displayed times */
  timeFormat: PropTypes.string.isRequired
};

TransitLegBody.defaultProps = {
  operator: null
};

export default TransitLegBody;
