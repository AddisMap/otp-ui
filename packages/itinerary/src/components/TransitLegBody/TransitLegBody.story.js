import React from "react";
import { storiesOf } from "@storybook/react";
import { withA11y } from "@storybook/addon-a11y";
import { withInfo } from "@storybook/addon-info";
import { action } from "@storybook/addon-actions";
import { withThemesProvider } from "storybook-addon-styled-component-theme";
import TransitLegBody from ".";
import itinerary from "../../__mocks__/itinerary";
import ThemeShim from "../../common/components/ThemeShim";
import MainStyleDecorator from "../../common/components/MainStyleDecorator";
import { LIGHT_THEME, DARK_THEME } from "../../common/constants";

const setActiveLeg = action("setActiveLeg");
const leg = itinerary.legs[1];
const legIndex = 1;
// Originally from OTP MOD config
const longDateFormat = "MMMM D, YYYY";
// Originally from OTP MOD config
const operator = {
  id: "TRIMET",
  name: "TriMet",
  logo:
    "http://news.trimet.org/wordpress/wp-content/uploads/2019/04/TriMet-logo-300x300.png"
};
// Originally from OTP MOD config
const timeFormat = "h:mm a";

storiesOf("Itinerary|Transit Leg Body", module)
  // .addDecorator(withThemesProvider([LIGHT_THEME, DARK_THEME]))
  .addDecorator(withA11y)
  .addDecorator(withInfo)
  .addDecorator(MainStyleDecorator)
  .add("Transit Leg Body", () => (
    <ThemeShim>
      <TransitLegBody
        leg={leg}
        legIndex={legIndex}
        setActiveLeg={setActiveLeg}
        longDateFormat={longDateFormat}
        operator={operator}
        timeFormat={timeFormat}
      />
    </ThemeShim>
  ));
