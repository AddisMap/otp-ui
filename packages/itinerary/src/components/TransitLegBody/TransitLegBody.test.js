import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import MOCK_THEME from "../../common/__mocks__/theme";
import TransitLegBody from ".";
import itinerary from "../../__mocks__/itinerary";

describe("Transit Leg Body Component", () => {
  const setActiveLeg = jest.fn(() => {});
  const leg = itinerary.legs[1];
  const legIndex = 1;
  // Originally from OTP MOD config
  const longDateFormat = "MMMM D, YYYY";
  // Originally from OTP MOD config
  const operator = {
    id: "TRIMET",
    name: "TriMet",
    logo:
      "http://news.trimet.org/wordpress/wp-content/uploads/2019/04/TriMet-logo-300x300.png"
  };
  // Originally from OTP MOD config
  const timeFormat = "h:mm a";

  let sut;

  beforeEach(() => {
    sut = render(
      <ThemeProvider theme={MOCK_THEME}>
        <TransitLegBody
          leg={leg}
          legIndex={legIndex}
          setActiveLeg={setActiveLeg}
          longDateFormat={longDateFormat}
          operator={operator}
          timeFormat={timeFormat}
        />
      </ThemeProvider>
    );
  });

  afterEach(() => jest.clearAllMocks());

  it("should match the snapshot", () => {
    const { asFragment } = sut;
    expect(asFragment()).toMatchSnapshot();
  });
});
