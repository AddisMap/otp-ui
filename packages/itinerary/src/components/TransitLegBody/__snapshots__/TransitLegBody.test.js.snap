// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transit Leg Body Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bwzfXH kGwkzX"
  >
    <button
      class="sc-bdVaJa sc-htpNat hncvTK"
    >
      <span
        class="sc-bxivhb hGsGIP"
      >
        Board 12-Barbur/Sandy Blvd 
        <span
          class="sc-ifAKCX khTzTx"
        >
          to 
        </span>
        Tigard TC via Portland City Center
      </span>
    </button>
    <div
      class="sc-htoDjs bZzhDu"
    >
      <button
        class="sc-bdVaJa sc-dnqmqq wOxjB"
      >
        23 min ride (17 stops) 
        <svg
          class="sc-iwsKbI htJanY"
          expanded="false"
          fill="#008ab0"
          height="10"
          mode="BUS"
          viewBox="0 0 24 24"
          width="10"
        >
          <path
            d="M0 7.33L2.829 4.5l9.175 9.339L21.171 4.5 24 7.33 12.004 19.5z"
          />
        </svg>
      </button>
      <div
        class="sc-gZMcBi liQYWD"
      >
        <div
          class="sc-gqjmRU hnZXkc"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
